File: src/agentx/discovery.py
from __future__ import annotations
import re
from pathlib import Path
from .config import BUCKETS, CONFIG_FILE, DISCOVERY_STEMS, ROOT_FALLBACK_FILES, Context, Paths

def _natural_key ( path: Path ) -> list:

    return [int(part) if part.isdigit() else part.lower() for part in re.split(r"(\d+)", path.name)]

def _config_in ( directory: Path ) -> Path | None:

    if not directory.is_dir():
        return None

    target = CONFIG_FILE.lower()

    for child in directory.iterdir():

        if child.is_file() and child.name.lower() == target:
            return child

    return None

def resolve_root ( start: Path ) -> Path:

    start = start.resolve()

    if _config_in(start):
        return start

    for parent in ( start, *start.parents ):

        if (parent / ".git").exists():
            return parent

    for parent in ( start, *start.parents ):

        if _config_in(parent):
            return parent

    return start

def _bucket_of ( stem: str ) -> str | None:

    for bucket, stems in DISCOVERY_STEMS.items():

        if stem in stems:
            return bucket

    return None

def discover ( paths: Paths ) -> Context:

    context = Context()

    if paths.docs.is_dir():

        for entry in sorted(paths.docs.iterdir()):

            if entry.is_file():

                if entry.suffix.lower() != ".md":
                    continue

                bucket = _bucket_of(entry.stem.lower())

                if bucket:
                    context.add(bucket, entry)

            elif entry.is_dir():

                bucket = _bucket_of(entry.name.lower())

                if bucket:

                    for md in entry.rglob("*.md"):

                        if md.is_file():
                            context.add(bucket, md)

    else:

        for child in sorted(paths.root.iterdir()):

            if child.is_file() and child.name.lower() in ROOT_FALLBACK_FILES:
                context.add("overview", child)

    for name in BUCKETS:
        context.bucket(name).sort(key=_natural_key)

    return context

File: src/app/discovery.rs
use std::path::{Path as StdPath, PathBuf};

use crate::config::Context;
use crate::config::Paths;
use crate::config::names::{CONFIG_FILE, ROOT_FALLBACK_FILES, bucket_of};
use crate::core::support::fs::{Dir, Path};
use crate::core::support::text::Text;

/// Resolve the project root: `Agentx.toml` in `start` -> nearest `.git` upward
/// -> nearest `Agentx.toml` upward -> `start`.
pub fn resolve_root ( start: &StdPath ) -> PathBuf {

    let start = start.canonicalize().unwrap_or_else(|_| start.to_path_buf());

    if has_config(&start) {
        return start;
    }

    for dir in start.ancestors() {

        if dir.join(".git").exists() {
            return dir.to_path_buf();
        }
    }

    for dir in start.ancestors() {

        if has_config(dir) {
            return dir.to_path_buf();
        }
    }

    start

}

/// Bucket the durable docs under `agents/` into the run context.
pub fn discover ( paths: &Paths ) -> Context {

    let mut context = Context::default();

    if paths.docs.is_dir() {

        for entry in Dir::entries(&paths.docs) {

            if entry.is_file() {

                if Path::has_extension(&entry, "md")
                    && let Some(bucket) = bucket_of(&Path::stem_of(&entry).to_ascii_lowercase())
                {
                    context.add(bucket, entry);
                }

            } else if entry.is_dir()
                && let Some(bucket) = bucket_of(&Path::name_of(&entry).to_ascii_lowercase())
            {
                let mut found = Vec::new();
                collect_markdown(&entry, &mut found);

                for md in found {
                    context.add(bucket, md);
                }
            }
        }

    } else {

        for entry in Dir::entries(&paths.root) {

            let name = Path::name_of(&entry).to_ascii_lowercase();

            if entry.is_file() && ROOT_FALLBACK_FILES.contains(&name.as_str()) {
                context.add("overview", entry);
            }
        }
    }

    for bucket in [
        &mut context.overview,
        &mut context.contracts,
        &mut context.history,
        &mut context.tasks,
        &mut context.requires,
    ] {
        bucket.sort_by(|a, b| Text::natural_compare(&Path::name_of(a), &Path::name_of(b)));
    }

    context

}

fn has_config ( dir: &StdPath ) -> bool {

    Dir::entries(dir).iter().any(|path| path.is_file() && Path::name_of(path).eq_ignore_ascii_case(CONFIG_FILE))

}

fn collect_markdown ( dir: &StdPath, out: &mut Vec<PathBuf> ) {

    for entry in Dir::entries(dir) {

        if entry.is_dir() {
            collect_markdown(&entry, out);
        } else if entry.is_file() && Path::has_extension(&entry, "md") {
            out.push(entry);
        }
    }

}
